src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -117,51 +114,40 @@ type Constraint a = SomePolynomial a
 
 -- | Adds a constraint to the arithmetic circuit.
 constraint :: Arithmetic a => Constraint a -> State (ArithmeticCircuit a) ()
-constraint con = modify $ \r -> r { acSystem = insert (con2var con) con (acSystem r) }
+constraint c = zoom #acSystem . modify $ insert (toVar [] c) c
 
 -- | Forces the current variable to be zero.
 forceZero :: forall a . Arithmetic a => State (ArithmeticCircuit a) ()
-forceZero = do
-    r <- get
-    let x   = acOutput r
-        con = polynomial [(one, monomial  (singleton x one))]
-    constraint con
+forceZero = zoom #acOutput get >>= constraint . var
 
 -- | Adds a new variable assignment to the arithmetic circuit.
 -- TODO: forbid reassignment of variables
-assignment :: forall a . (Map Integer a -> a) -> State (ArithmeticCircuit a) ()
-assignment f = modify $ \r -> r { acWitness = (insert (acOutput r) =<< f) . acWitness r }
+assignment :: (Map Integer a -> a) -> State (ArithmeticCircuit a) ()
+assignment f = do
+    i <- insert <$> zoom #acOutput get
+    zoom #acWitness . modify $ (.) (f >>= i)
 
 -- | Adds a new input variable to the arithmetic circuit. Returns a copy of the arithmetic circuit with this variable as output.
 input :: forall a . State (ArithmeticCircuit a) (ArithmeticCircuit a)
-input = modify (\(r :: ArithmeticCircuit a) ->
-        let ins    = acInput r
-            s      = if null ins then 1 else maximum (acInput r) + 1
-        in r
-        {
-            acInput    = ins ++ [s],
-            acOutput   = s,
-            acVarOrder = singleton (0, s) s
-        }) >> get
+input = do
+  inputs <- zoom #acInput get
+  let s = if null inputs then 1 else maximum inputs + 1
+  zoom #acInput $ modify (++ [s])
+  zoom #acOutput $ put s
+  zoom #acVarOrder $ put $ singleton (0, s) s
+  get
 
 -- | Evaluates the arithmetic circuit using the supplied input map.
 eval :: ArithmeticCircuit a -> Map Integer a -> a
-eval ctx i =
-    let w = acWitness ctx i
-        o = acOutput ctx
-    in w ! o
+eval ctx i = acWitness ctx i ! acOutput ctx
 
 -- | Applies the values of the first `n` inputs to the arithmetic circuit.
 -- TODO: make this safe
 apply :: [a] -> State (ArithmeticCircuit a) ()
-apply xs = modify (\(r :: ArithmeticCircuit a) ->
-    let ins = acInput r
-        n   = length xs
-    in r
-    {
-        acInput = drop n ins,
-        acWitness = acWitness r . (fromList (zip (take n ins) xs) `union`)
-    })
+apply xs = do
+    inputs <- zoom #acInput get
+    zoom #acInput . put $ drop (length xs) inputs
+    zoom #acWitness . modify $ (. (union . fromList $ zip inputs xs))
```

Comment:
Flipping `union` arguments would make it a little more readable: `(. union (fromList $ zip inputs xs))`